Add support to `Tensor[]?` for structured kernel codegen.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69607
* __->__ #69606

This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc @ezyang @bhosmer @bdhirsh

ysiraichi committed:
Add support to `Tensor[]?` for structured kernel codegen.

This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/core/IList.h:
```diff
@@ -0,0 +1,258 @@
+ * Interface that implements key functions for each `IListTag` type.
+ *
+ * You should create an specialization of this class for each
+ * possible combination of `IListTag` type (except `None`) and
```
Maybe just enumerate the current full list here, for clarity:
`(at::Tensor, IListTag::unboxed)`
`(at::Tensor, IListTag::boxed)`
`(at::OptionalTensorRef, IListTag::unboxed)`
`(at::OptionalTensorRef, IListTag::boxed)`

bdhirsh commented on aten/src/ATen/core/IList_inl.h:
```diff
@@ -0,0 +1,215 @@
+// [Note: ITensorList]
+using ITensorList = c10::IList<at::Tensor>;
+using ITensorListIterator = c10::IListIterator<at::Tensor>;
+// [Note: at::OptionalTensorRefRefList]
```
Nit: this doesn't match `[Note: IOptTensorRefList]` in the codegen below

ysiraichi replied:
Ah, you are right. Nice catch!

bdhirsh commented on aten/src/ATen/core/IList_inl.h:
```diff
@@ -0,0 +1,215 @@
+}
+
+namespace at {
+// [Note: ITensorList]
```
Are you going to add a description here of why we have / need ITensorlist (and IOptTensorRefList)? Maybe summarizing some of the discussion from the gh issue so people can quickly come to this file and understand why we have these two new types.

Separately - you have a bunch of good comments defined in `IList.h`. Can you add a bit more on (a) how to add a new tensor type (Tensor, OptionalTensorRef, _____), and (b) how to add a new tag (IListTag::unboxed, IListTag::boxed, _____)? For example, it looks like we need to define a new `IListTagImpl` for the full cross product of [tags] x [tensor types]

ysiraichi replied:
> Are you going to add a description here of why we have / need ITensorlist (and IOptTensorRefList)?

I wasn't planning to do that. But, I guess it would be better to do so.

> Can you add a bit more on (a) how to add a new tensor type (Tensor, OptionalTensorRef, _____), and (b) how to add a new tag (IListTag::unboxed, IListTag::boxed, _____)? For example, it looks like we need to define a new `IListTagImpl` for the full cross product of [tags] x [tensor types]

Sure, I will do that. Yeah, currently we need to do that, which is quite verbose. The reason for that is that `list_type` for the boxed and unboxed versions might change, depending on what is the parameter type of `IList` (see below).

```c++
/* IList<at::Tensor> */
IListTagImpl<at::Tensor, IListTag::Unboxed>::list_type == ArrayRef<at::Tensor>
IListTagImpl<at::Tensor, IListTag::Boxed>::list_type == List<at::Tensor>

/* IList<at::OptionalTensorRef> */
IListTagImpl<at::OptionalTensorRef, IListTag::Unboxed>::list_type == ArrayRef<at::OptionalTensorRef>
IListTagImpl<at::OptionalTensorRef, IListTag::Boxed>::list_type == List<optional<at::Tensor>>
```

That said, I'll try to play with it a little bit more and see if I can get a better design.

bdhirsh commented on aten/src/ATen/core/IList_test.cpp:
```diff
@@ -1,28 +1,52 @@
 using namespace c10;
 
 static std::vector<at::Tensor> get_tensor_vector() {
   std::vector<at::Tensor> boxed;
```
nit: `boxed` -> `tensors`?

ysiraichi replied:
Better, indeed.

bdhirsh commented:
I think overall this looks pretty good :)

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi commented:
Hey, @bdhirsh
Thank you for your feedback. Here's a summary of the changes I have introduced:

- Tried to reduce boilerplate when supporting a new type (I feel like it's way better now)
    - Created a class with the default implementations: `IListTagImplBase<ListT, T, TAG>`
    - So, for implementing `IListTagImpl<T, TAG>`, it's as simple as inheriting from `IListTagImplBase<SomeListT, T, TAG>` and re-implementing the methods that differ
    - Good, because the `unwrap` methods should be the same, most of the time
- Added more comments for clarifying how each of the templated classes interact with each other
- Added comments for (as suggested):
    - How to add a new `IListTag`
    - How to add support for a new type
- Removed one of the macros (`TORCH_ILIST_IMPL`), favoring clarity 

Have a look, and let me know what you think!

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi commented:
I have changed bits of this code for easing the adoption of `IList` in the dispatcher. Here's a summary:

- Changed the order of the template parameters for `IListTagImpl{,Base}` (the `IListTag` value comes first) like:
    - `IListTagImpl<TAG, T>`
    - `IListTagImplBase<TAG, T, ListElemT>`
- `IListTagImplBase` now defines a default `list_type` over `elem_type`
    - Instead of varying the whole `list_type` for each `(TAG, type)`, we vary the element it holds: `elem_type`
    - Now, `Boxed` (`Unboxed`) tag assumes that the container is a `List<elem_type>` (`ArrayRef<elem_type>`)
    - **Why?** It probably makes `IList<T>` implementation easier to understand
- `IListTagImpl<Boxed, OptionalTensorRef>::iterator_get` implementation was moved to `IList.cpp` for avoiding the inclusion of `Tensor.h` in `IList.h`.

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Fix a few compiler errors. on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Remove stale comment. on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Rebased to viable/strict. on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Invoke constructor explicitly when translating. on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ezyang commented:
this is still not ready for review, is that right?

ysiraichi commented:
Ah. Forgot to flag it as "ready". Sorry.

ysiraichi committed:
Rebased to 'viable/strict.' on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/IList.h:
```diff
@@ -0,0 +1,512 @@
+#pragma once
```
No changes needed here, but this is the sort of thing where I don't think stacking the PRs helps much, it would be better to merge the PRs into one mega PR so I can see where everything is going to be finally.

ysiraichi replied:
Hmmm. Is it because they are incremental PRs that overlap (on the code they are changing)? I thought this was exactly the type of changes where it made sense using a stack of PRs, since:
- There are less code to be reviewed at a time
- Makes it easier to merge and revert

On the other hand, it does feel like the code is always changing, though. But I thought the benefits outweighed the drawbacks. Anyway, I will try to think a bit more next time.

ezyang replied:
It's definitely a delicate balance. On the one hand reducing the amount of code to review makes review easier. But if you ask me if I'd prefer one PR (200LOC), or two PRs (150LOC each) where the first PR takes a detour before getting to the final destination in the second PR, I'd prefer one PR, it's less work net for me and I don't have to spend time understanding an intermediate state that will immediately get moved away from.

That's not to say that intermediate step PRs don't have their place; for example when a change is too complex/risky to do in one go, taking detours is a good way to derisk. But if you introduce something new and then immediately change it next, well, that just suggests the diff stack is just in chronological order rather than ease for review. And sometimes it is possible to split up the diff for review, but it's not worth it (I'm allowed to say this, cuz I'm the reviewer :P ).

ysiraichi replied:
That does make sense. I will make sure to take that into consideration next time.
Anyway, do you think I should I merge #67964 and #69606?

ezyang replied:
It's up to you, we've reviewed these PRs now so it is mostly a matter of how to get these over the finish line one way or another.

ezyang commented on aten/src/ATen/core/IList.h:
```diff
@@ -0,0 +1,512 @@
+ * - `IList<at::Tensor>`
+ * - `IList<at::OptionalTensorRef>`
+ *
+ * Note that `ITensorList` is a view type. Meaning that it won't own the
```
IList

ezyang commented on aten/src/ATen/core/IList.h:
```diff
@@ -0,0 +1,512 @@
+ * Note that `ITensorList` is a view type. Meaning that it won't own the
+ * tensors it holds. If you need it to last longer, make sure that there is
+ * actually a non-temporary list of tensors (e.g. `vector<Tensor>`) that owns
+ * them and outlives the `ITensorList` instance.
```
It's probably better to just say that it is intended to be used only for argument parameters (where temporaries are guaranteed to live long enough because they will live for the entirety of the function call)

ezyang commented on aten/src/ATen/core/IList.h:
```diff
@@ -0,0 +1,512 @@
+ * (currently, only boxed and unboxed), without incurring in extra
+ * overhead for converting from one to another. It does so while
+ * exposing usual container methods, which dispatch to corresponding
+ * implementations.
```
Maybe it would be good to say that the intended use of this interface is iterating over it once (and not repeatedly accessing elements from the container).

ezyang commented on aten/src/ATen/core/IList.h:
```diff
@@ -0,0 +1,512 @@
+  /*
+   * For convenience, enable other supported `unboxed_type` constructors as
+   * explicit constructors.
+   */
```
Hmm, I'm confused. Don't you need to support *all* implicit constructors from ArrayRef for IList to be a drop-in replacement for ArrayRef?

ysiraichi replied:
Yeah. This came before we discussed in the other PR. I will make them implicit.

ezyang commented:
Can I pass an ArrayRef<Tensor> to an IOptTensorList? Ordinarily this would require a conversion and it would be nice if it wasn't required (because it means I can now say `{tensor, tensor2}` and have it work out even if it's an IOptTensorList...)

ysiraichi commented:
Not yet. But I will try to work on it.

ysiraichi committed:
Unmerge `cat` and `Tensor?[]` support (mistakenly removed). on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34706812).

ezyang committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ezyang commented:
I rebased this stack myself cuz it was merge conflicting

ezyang commented on aten/src/ATen/NamedTensorUtils.cpp:
```diff
@@ -459,7 +459,7 @@ std::vector<Dimname> broadcast_to_outnames(
 }
 
-std::vector<Dimname> compute_cat_outnames(ITensorListRef tensors) {
+std::vector<Dimname> compute_cat_outnames(ITensorList tensors) {
```
Hmm, this doesn't look like it was supposed to happen...

ysiraichi replied:
Ah, I thought it would be easier to keep the old names in these PRs. Clearly, I was wrong.

ezyang commented:
Hmm, I may have made a hash of this, you may need to just force push your own local rebase; I see the Ref changes get flipped back in this PR and that doesn't look appropriate.

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/IListRef.h:
```diff
@@ -0,0 +1,612 @@
+#include <type_traits>
+
+/*
+ * Adding support for a new Tag
```
nit: this comment is kind of in the wrong place, it'd be better to see the high level comment for what an IListRef is first haha

ezyang commented on aten/src/ATen/core/IListRef.h:
```diff
@@ -0,0 +1,612 @@
+ * ===========================
+ * MSVC `vector<T>::iterator` implementation (used in the boxed variant)
+ * makes it so this union's destructor, copy-constructor (assignment), and
+ * move-constructor (assignment) are implcitly deleted.
```
nit: implicitly

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ezyang commented:
test failures are valid

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge this

suo commented:
@pytorchbot revert this

This PR looks like it broke a windows libtorch debug build: https://github.com/pytorch/pytorch/runs/6040603959?check_suite_focus=true

malfet commented:
Reverting it manually, but we need to figure out the workflow

ezyang commented:
@malfet Probably because it is doing some stuff with C++ iterators, and on Windows how exactly iterators are defined vary depending on if you are DEBUG or not

ezyang commented:
error is now in pr ci

```
C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\aten\src\ATen/core/IListRef.h(381): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\aten\src\ATen/core/IListRef.h(475): note: see reference to class template instantiation 'c10::IListRefIterator<T>' being compiled
C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\aten\src\ATen/core/IListRef.h(381): error C2143: syntax error: missing ',' before '&'
```

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ysiraichi commented:
Ah, my bad. I forgot to rename from `ITensorListRefIterator` to `IListRefIterator` in the `IListRefIterator` copy-constructor. I believe it should be fixed, now.

ezyang commented:
still failing

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ysiraichi commented:
My bad. I forgot to inspect the body of the constructor, too.
@ezyang Could you try again?

ezyang committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ezyang committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

Differential Revision: [D34706812](https://our.internmc.facebook.com/intern/diff/D34706812)

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge this

ezyang commented:
@pytorchbot merge this

ezyang committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ysiraichi committed:
Update on "Add support to `Tensor[]?` for structured kernel codegen."


This PR turns the previously introduced `ITensorList` into a more general `IList`
class. It is a container wrapper for arbitrary types (given their appropriate
implementations).

In summary, I have:

- Renamed `ITensorList` (its iterators and macros, for consistency) to `IList`
- Made `IList` a templated function (for an arbitrary type `T`), given that they:
     - Specialize `IListTagImpl<T, Tag>`, for all `IListTag`
- Introduced type aliases (for both list and iterator types):
     - `at::ITensorList` -> `c10::IList<at::Tensor>`
     - `at::IOptTensorRefList` -> `c10::IList<at::OptionalTensorRef>`
- Added support for `Tensor?[]` in the structured codegen

cc ezyang bhosmer bdhirsh

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge this
